src/algorithms/hnsw.rs

```diff
@@ -509,6 +531,55 @@ pub fn local_search(
     results
 }
 
+fn local_search_vbase<'mmap, 'vector>(
+    mmap: &'mmap HnswMmap,
+    range: usize,
+    s: u32,
+    vector: &'vector [Scalar],
+) -> HnswIndexIter<'mmap, 'vector> {
+    assert!(range > 0);
+    let mut visited_guard = mmap.visited.fetch();
+    let mut visited = visited_guard.fetch();
+    let mut candidates = BinaryHeap::<Reverse<(Scalar, u32)>>::new();
+    let mut results = Heap::new(range);
+    visited.mark(s);
+    let s_dis = mmap.quantization.distance(mmap.d, vector, s);
+    candidates.push(Reverse((s_dis, s)));
+    results.push(HeapElement {
+        distance: s_dis,
+        payload: mmap.raw.payload(s),
+    });
+    while let Some(Reverse((u_dis, u))) = candidates.pop() {
+        if !results.check(u_dis) {
+            break;
+        }
+        let edges = find_edges(mmap, u, 0);
+        for &HnswMmapEdge(_, v) in edges.iter() {
+            if !visited.check(v) {
+                continue;
+            }
+            visited.mark(v);
+            let v_dis = mmap.quantization.distance(mmap.d, vector, v);
+            if !results.check(v_dis) {
+                continue;
+            }
+            candidates.push(Reverse((v_dis, v)));
+            results.push(HeapElement {
```

Comment:
Will it lost data to be returned?

Reply:
Yes.

Reply:
The first stage points should be saved and sorted. In the second stage, we will return a better choice in first stage points and second stage heap.

Reply:
Not all, at most `range` points here.

Reply:
The line 578 saves the the first stage heap to the iterator. The current implementation has already chosen the first stage point.

Reply:
First stage heap contains at most `range` points. All points in first stage, including points that are not in first stage heap, should be returned.

Reply:
What do you mean by the points in the first stage but not in the first stage heap?

Reply:
All visited points. `results` will contain at most `range` points, and points are deleted by `heap::push` are lost.